sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_arg_restrict_property.asciidoc

```diff
@@ -0,0 +1,187 @@
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+== Contributors
+
+Abhishek Tiwari, Intel +
+Joseph Garvey, Intel
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, revision 4.
+
+It depends on the following extensions:
+
+ - link:../experimental/sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
+ - link:sycl_ext_oneapi_annotated_ptr.asciidoc[sycl_ext_oneapi_annotated_ptr]
+ - link:sycl_ext_oneapi_annotated_arg.asciidoc[sycl_ext_oneapi_annotated_arg]
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+== Overview
+
+This extension introduces the `restrict` kernel argument property for the
+classes `sycl::ext::oneapi::annotated_ptr` and
+`sycl::ext::oneapi::annotated_arg`.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_KERNEL_ARG_RESTRICT_PROPERTY` to one of the values
+defined in the table below.  Applications can test for the existence of this
+macro to determine if the implementation supports this feature, or applications
+can test the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== `restrict` property
+
+The `restrict` property defined below is supported with the
+`sycl::ext::oneapi::experimental::annotated_ptr` and
+`sycl::ext::oneapi::experimental::annotated_arg` classes. They are only
+meaningful on the kernel argument when the kernel argument is a pointer type or
+a pointer wrapper type and are not meaningful within the kernel body.
+
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+struct restrict_key {
+  using value_t = property_value<restrict_key>;
+};
+
+inline constexpr restrict_key::value_t restrict;
+
+template<>
+struct is_property_key<restrict_key> : std::true_type {};
+
+template<typename T, typename PropertyListT>
+struct is_property_key_of<
+  restrict_key, annotated_ptr<T, PropertyListT>> : std::true_type {};
+
+template<typename T, typename PropertyListT>
+struct is_property_key_of<
+  restrict_key, annotated_arg<T, PropertyListT>> : std::true_type {};
+} // namespace sycl::ext::oneapi::experimental
+```
+--
+
+[frame="topbot",options="header"]
+|===
+|Property |Description
+
+a|
+[source,c++]
+----
+restrict
+----
+a|
+This is a hint to the compiler that the pointer kernel arguments marked with
+this property do not alias with one another.
```

Comment:
Updated to state that.

Reply:
Sorry to be pedantic, but I think we should use the word "assert" instead of "hint" and we should state that it is UB if the pointers do in fact alias with one another.

For example, I think the wording could be like:

> This is an assertion by the application that the pointer kernel arguments marked with this property do not alias with one another with the same semantics as the C99 `restrict` keyword.  The behavior is undefined if these pointer values do alias.

Reply:
Please don't apologize, it's good to be rigorous. Updated :)